### ISSUE TITLE
inodes not freed after repeated builds - aufs

### ISSUE THREAD

**Issue by mvording** [moby/moby#9755]
Description of problem:

For this scenario, docker build is being used as part of a CI workflow.   After several months of builds, even with having a daily container cleanup script in place, this failed due to being out of space.   Checking the disk,  there was plenty of disk space available, but no inodes remaining.   Checking the filesystem,  it turns out the /usr/lib/docker folder tree contained several million inodes.

docker rm/rmi commands were used to remove all containers and images from the system.   However this only freed up a few % of the inodes.

To resolve this I had to manually do a "rm -rf /var/lib/docker" to free the inodes.

`docker version`:
Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3.3
Git commit (client): 5bc2ff8
OS/Arch (client): linux/amd64
Server version: 1.4.1
Server API version: 1.16
Go version (server): go1.3.3
Git commit (server): 5bc2ff8

`uname -a`:
Linux ip-10-1-0-63 3.13.0-32-generic #57-Ubuntu SMP Tue Jul 15 03:51:08 UTC 2014 x86_64 x86_64 x86_64 GNU/Linux

Environment details (AWS, VirtualBox, physical, etc.):
AWS

How reproducible:
Run many docker builds with differing source files, watch the inode count keep increasing and not get reset significantly when all containers and images are removed.

**Comment by donpinkster** [comment]
We also just encountered this, nuking /var/lib/docker is the easiest solution. But it would be great if this was not needed of course.

**Comment by cpuguy83** [comment]
Question: Is `/var/lib/docker/aufs` empty after removing all containers/images?

**Comment by sebbacon** [comment]
Given the days of `aufs` [appear numbered](http://developerblog.redhat.com/2014/09/30/overview-storage-scalability-docker/), we'll try the `DeviceMapper` backend and see if that suffers from the same problem, and report back here.

**Comment by sebbacon** [comment]
We didn't get very far with the `DeviceMapper` backend, because issue #4036 prevented us from building any new images.  Reverting to aufs with a larger inode table (and nuking `/var/lib/docker/aufs`)  for the time being.

**Comment by morgante** [comment]
We (@efuquen and I) are having this same issue. Even after a full system wipe, Docker quickly uses up all inodes after building even a few images.

**Comment by efuquen** [comment]
This is my inode usage based off the helpful script @vincentwoo provided:

```
------------------------------------------
	INODE USAGE SUMMARY
------------------------------------------
 INODES    |  SIZE      | DIRECTORY
------------------------------------------
 67        |  544K      | containers
 13        |  52K       | execdriver
 440       |  1.8M      | graph
 2         |  7.5M      | init
 3878564   |  3.6G      | overlay
 1         |  4.0K      | tmp
 1         |  4.0K      | trust
 12        |  44K       | vfs
 13        |  52K       | volumes
------------------------------------------
3957193    | 3.6G       | /var/lib/docker
------------------------------------------
```

**Comment by sebbacon** [comment]
Quick update:  using the `btrfs` backend is a successful workaround for us

**Comment by morgante** [comment]
@sebbacon Are you on CoreOS?

**Comment by mrcoles** [comment]
I just ran into this on an EC2 machine

**Comment by frankfuu** [comment]
Hi @vincentwoo ,

Just wondering if you could provide me with the helpful script that was given to @efuquen ?

**Comment by vincentwoo** [comment]
@frankfuu it was just https://github.com/tripflex/inodes

**Comment by frankfuu** [comment]
Thanks @vincentwoo 

I'm still having the same issues, and after a few builds I need stop docker then  nuke ```/var/lib/docker``` directory then restart ```docker``` and rebuild my apps on dokku and restart ```mongod``` before  working again

**Comment by Jaykah** [comment]
Have the same issue with overlay:

```
root@Logstash1:/jenkins/docker# find . -xdev -type f | cut -d "/" -f 2 | sort | uniq -c | sort -n
      1 init
      1 linkgraph.db
      1 repositories-overlay
      1 trust
     30 containers
    116 graph
    196 tmp
1445877 overlay
```

docker info
```
Containers: 5
Images: 56
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 3.18.0-031800-generic
Operating System: Ubuntu Wily Werewolf (development branch) (containerized)
CPUs: 4
Total Memory: 6.805 GiB
Name: Logstash1
ID: MYDD:REBI:AB6H:RLW3:YZD2:SUWT:CJED:ENVW:7W3U:RYUC:ZMM4:R2DI
WARNING: No swap limit support
```

docker version
```
Client version: 1.7.0
Client API version: 1.19
Go version (client): go1.4.2
Git commit (client): 0baf609
OS/Arch (client): linux/amd64
Server version: 1.7.0
Server API version: 1.19
Go version (server): go1.4.2
Git commit (server): 0baf609
OS/Arch (server): linux/amd64
```


DIND is being used for our CI workflow.

**Comment by Jaykah** [comment]
Ping https://github.com/docker/docker/pull/12354

**Comment by FelikZ** [comment]
Almost a year since issue exists :(

**Comment by cpuguy83** [comment]
Let's not mix overlay issues with aufs issues please.
Overlay has known issues with excessive inode usage.

I have not personally had any issues with excessive inode usage on aufs, but have seen it be flakey (because aufs can be flakey) when removing containers/images (ie, not actually removing them, docker does report the error).
Are you getting removal errors on your transient containers?

**Comment by thefallentree** [comment]
We are seeing this issue too, although, not from build, but from excessive create/delete script. 

Anyone have any idea how to proceed, we are running out of inodes every couple days, and it is painful to keep things up.

**Comment by zbyte64** [comment]
Just experienced this. Given enough deploys in production it seems unavoidable.

**Comment by mrcoles** [comment]
Is there an updated analysis of what’s going wrong with this problem? The initial post said that running `docker rmi` didn’t help as much as it should? Does this affect machines that build and not ones that just pull?

I’m using AWS linux ami machines, and for my production machines, which only pull, I added a post deploy script to at least run `docker rmi $(docker images -q)`, but for my jenkins build machine, it seems I just have to ssh in and perform the more drastic steps to clean it up when things go south.

**Comment by cpuguy83** [comment]
I wonder how many volumes you have on the host and if this is the source of your inode usage.

**Comment by zbyte64** [comment]
I run `docker rmi $(docker images -q --filter "dangling=true")` after each deploy but it did not free up nearly enough inodes. Ended up wiping the entire aufs folder and rebuilding from scratch.

**Comment by FelikZ** [comment]
@zbyte64  +1

**Comment by yoanisgil** [comment]
Ran into this issue as well on my development environment (haven't hit this yet on production but we're only running a few contianers (~ 12)). I heavily use docker-compose for setting up a dev stack so I do run a lot of docker-compose up/run. I must say I've been using docker for dev purposes on a day to day basis for 3 - 4 monts now, so itś not the same as what @thefallentree described. 

I have no choice but to nuke /var/lib/docker, so I will do that and keep an eye on this issue. What worries me the most is the impact this issue might have on a production environment. I was wondering if others could share their FS setups and how good (or bad has worked for them). Right now we're using XFS with LVM but as I said before we're only running around ~ 12 containers and there are roughly 2 deploys/day.

 Here is some information:

```bash
# uname -a
Linux aws015 4.1.6-200.fc22.x86_64 #1 SMP Mon Aug 17 19:54:31 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
# docker info
Containers: 3
Images: 619
Storage Driver: overlay
 Backing Filesystem: extfs
Execution Driver: native-0.2
Logging Driver: syslog
Kernel Version: 4.1.6-200.fc22.x86_64
Operating System: Fedora 22 (Twenty Two)
CPUs: 4
Total Memory: 15.5 GiB
Name: aws015
ID: UHNK:E7RB:24OQ:FQ2G:ZNOW:HSTJ:CDOQ:FZET:EXTT:SJJH:AGHR:5GXB
# docker version
Client:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:39:27 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.8.1
 API version:  1.20
 Go version:   go1.4.2
 Git commit:   d12ea79
 Built:        Thu Aug 13 02:39:27 UTC 2015
 OS/Arch:      linux/amd64
# df -i /
Filesystem                      Inodes   IUsed IFree IUse% Mounted on
/dev/mapper/fedora_aws015-root 3276800 3275073  1727  100% /
# df -h /
Filesystem                      Size  Used Avail Use% Mounted on
/dev/mapper/fedora_aws015-root   50G   28G   20G  59% /
#tune2fs -l /dev/mapper/fedora_aws015-root
tune2fs 1.42.12 (29-Aug-2014)
Filesystem volume name:   <none>
Last mounted on:          /
Filesystem UUID:          13e67bf8-0279-4514-97ea-c788b0fd634f
Filesystem magic number:  0xEF53
Filesystem revision #:    1 (dynamic)
Filesystem features:      has_journal ext_attr resize_inode dir_index filetype needs_recovery extent flex_bg sparse_super large_file huge_file uninit_bg dir_nlink extra_isize
Filesystem flags:         signed_directory_hash 
Default mount options:    user_xattr acl
Filesystem state:         clean
Errors behavior:          Continue
Filesystem OS type:       Linux
Inode count:              3276800
Block count:              13107200
Reserved block count:     655360
Free blocks:              5641750
Free inodes:              1202
First block:              0
Block size:               4096
Fragment size:            4096
Reserved GDT blocks:      1020
Blocks per group:         32768
Fragments per group:      32768
Inodes per group:         8192
Inode blocks per group:   512
Flex block group size:    16
Filesystem created:       Thu Jul  2 15:58:50 2015
Last mount time:          Fri Sep 11 15:28:49 2015
Last write time:          Fri Sep 11 11:28:45 2015
Mount count:              10
Maximum mount count:      -1
Last checked:             Thu Jul  2 15:58:50 2015
Check interval:           0 (<none>)
Lifetime writes:          232 GB
Reserved blocks uid:      0 (user root)
Reserved blocks gid:      0 (group root)
First inode:              11
Inode size:	          256
Required extra isize:     28
Desired extra isize:      28
Journal inode:            8
First orphan inode:       2623233
Default directory hash:   half_md4
Directory Hash Seed:      42714f67-ce1f-4425-9693-6ab291abde2e
Journal backup:           inode blocks
```

**Comment by sebbacon** [comment]
[`btrfs` continues to work fine for us](https://github.com/docker/docker/issues/9755#issuecomment-73356075) @yoanisgil

**Comment by wouafff** [comment]
We have the same issue with aufs.
rmi doesn't change anything, we have to nuke /var/lib/docker.

ps: Backing Filesystem: extfs

**Comment by elyulka** [comment]
I have the same problem with overlay+ext4.

**Comment by cpuguy83** [comment]
@elyulka this is a known issue with overlay fs.

**Comment by cristianocasella** [comment]
Hi, same problem here -.-

**Comment by cristianocasella** [comment]
Hi guys, if you're in trouble give a look to this post:
http://masato.github.io/2014/11/05/how-to-correctly-remove-volume-directories/

or put in cron this command: (it's stupid but it works)
docker ps -a | grep Exited | awk '{ print $1 }' | xargs docker rm --volumes=true

**Comment by thaJeztah** [comment]
@cristianocasella removing volumes is *not* related here, also docker 1.9 has a `docker volumes` command to manage volumes. The script you're proposing, is better written using the built-in filter functionality of docker; i.e. `docker rm -v $(docker ps -q -f status=exited)`

**Comment by Bendihossan** [comment]
Related to #10613

**Comment by F21** [comment]
Also thought I would chime in here. I am running docker 1.9.1 with aufs with docker-compose 1.5.2 on Ubuntu 15.10 64-bit in a VMWare Workstation VM on Windows 8.1.

I am currently prototyping a HDFS HA and HBase HA cluster using docker. Using docker-compose, I launch 13 containers and 1 data volume container.

I have noticed that with a clean VM at the start of the day, things work pretty well. However, by early afternoon, containers won't stop and a lot of times, the HBase master and regionserver processes won't start or run correctly in the containers. I can see that my entrypoint is run and it starts the process for the master or regionserver, but they no longer work properly or produce any output. I also noticed that Ubuntu will complain about a lot of orphan inodes when I try to restart the VM.

The only way to fix this problem is to delete everything in `/var/lib/docker` and start the docker daemon again.

**Comment by andreychernih** [comment]
I am also having this problem with default docker machine in Mac OS X. I have to nuke the whole `/var/lib/docker` to fix this. One interesting thing was that `docker info` was showing 500 images, but when I did `docker images|wc -l`, it only showed 150.

**Comment by thaJeztah** [comment]
@andreychernih the difference in images-count is likely because `docker info` also includes the number of *intermediate* images, i.e. `docker images -aq | wc -l` should match that number more closely

**Comment by andreychernih** [comment]
@thaJeztah thanks, with intermediate images shown, these numbers are closer indeed.

**Comment by sandys** [comment]
hi guys, I just hit the same inode exhaustion issue with overlayfs. What is the best practice around this ? We do multiple deploys a day and this problem is hitting us hard.

**Comment by thaJeztah** [comment]
@sandys the [overlay section](https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/#overlayfs-and-docker-performance) in the documentation has some hints, although in general, it's a result of the way overlay works.

**Comment by szimszon** [comment]
I'm having the same issue. :(

**Comment by mfornasa** [comment]
+1

**Comment by lox** [comment]
We run into this issue nearly daily now with `overlayfs`. Because of docker-machine, the only fix is essentially to destroy our machines and start again. Is this the canonical bug for overlayfs inode exhaustion @thaJeztah? Should I be filing one specifically with docker-machine?

The only other reference I could find to this issue is https://github.com/coreos/bugs/issues/264

**Comment by thaJeztah** [comment]
@lox the issue tracking it in docker/docker is https://github.com/docker/docker/issues/10613, and you can partly workaround the issue e.g. https://github.com/docker/docker/issues/10613#issuecomment-122182216, https://docs.docker.com/engine/userguide/storagedriver/overlayfs-driver/#overlayfs-and-docker-performance). Unfortunately, exhaustive node usage is due to the way overlay works, so making those configuration changes (for now) is the best workaround. Perhaps docker-machine could make those changes when creating a new machine, yes

**Comment by HackToday** [comment]
hi @thaJeztah  Could you help explain why overlay have inode limitation, From the overlay point of view, it also create directories and files (some with hard link, which would not increase inode). Other storage driver also create many directiese and files, I did not quite understood it.

**Comment by saada** [comment]
We are running into this issue using OverlayFS with an inode limit of: 3276800. That's over 3 Million inodes. Shouldn't that cover enough ground for a dozen builds/runs per day?
How is this still a problem since Dec 19, 2014?!?!?

**Comment by cpuguy83** [comment]
@saada the issue with overlay is not the same and has a well known cause.
It's not that inodes are not released, it's that you have images laying around and the overlay driver used a lot of inodes.

**Comment by lox** [comment]
Is the implication then @cpuguy83 that it's possible to reclaim those inodes by purging images? I was under the impression that wasn't possible.

**Comment by cpuguy83** [comment]
@lox again, the issue of overlay and this issue are not the same at all.

I ___suspect___ the issue with aufs is images (or the container) that got something like a "device or resource busy" error when trying to remove, but no longer show up in `docker images` or `docker ps -a` because the configurations are gone, but the data is still there.
Issues with "device or resource busy" should be pretty much all fixed in 1.11.

**Comment by jdunmore** [comment]
I can run one of the various docker clean up scripts that helps - but if I
really want to clean it up I have to stop docker rm -rf /var/lib/docker
then I get the iNodes back.
Specifically, mine is related to running gitlab runners on coreos.

**Comment by lox** [comment]
Ah yup, you are correct apologies @cpuguy83, I hadn't realized the initial report was AUFS based.

**Comment by gmistick** [comment]
We currently run docker images with aufs, I'm creating and deleting containers all the time.
The HDD (/var/lib/docker/aufs/diff) is 899GB in a matter of few days....

Any fix for this / recommended storage-driver?

**Comment by thaJeztah** [comment]
@gmistick could you include output of `docker info` and `docker version`? Also what's running inside those containers? Are you using `docker exec` to start additional processes in the container that may keep mounts in use?

**Comment by gmistick** [comment]
@thaJeztah 
0. Container churn can be somewhat high. (at least 10 containers removed and restored daily).
1. Not running much exec.
2. It may happen that  a container takes too long to start ( internal app) and so the inner app churn but the inner app external tools ( processes) are getting killed with forceful commands ( "kill -9 .. external tools)
3. Volumes are always mapped to local disks and are written to quite a lot ( logs ). (It allows for persistence when deleting and recreating the conatiner)
4. Containers can get deleted ( rm ) and restored many times a day.



Docker Info and version:

Containers: 75
 Running: 10
 Paused: 0
 Stopped: 65
Images: 3
Server Version: 1.12.1
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-31-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 48
Total Memory: 62.81 GiB
Name: dabor
ID: XQZO:4PSG:LRIY:37XP:73AJ:LKDK:TJQQ:PMF3:PRCK:Y7LX:S5B4:6WTO
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false


Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 05:33:38 2016
 OS/Arch:      linux/amd64

**Comment by thaJeztah** [comment]
Thanks for the additional information @gmistick, it may also be worth checking the daemon logs, to see if they contain a hint as to why things are left behind.

When you say "kill -9 .. external tools"; what exactly do you mean with "external tools"?

also /cc @anusha-ragunathan perhaps you have more things to ask

**Comment by thaJeztah** [comment]
@gmistick if you remove all containers, and images is the disk space returned? Is all space taken inside the `aufs` directory, or could this be volumes that were not removed?

**Comment by anusha-ragunathan** [comment]
@thaJeztah : Nothing in here screams `EBUSY on aufs during container remove`. This is tracking growing inodes issues. Maybe @mlaventure data management commands can help?

**Comment by gmistick** [comment]
@thaJeztah  - All disk space is returned if i remove all containers and images. 
yes- all space is taken by the diffs directory inside the aufs directory.

Use case for forceful shutdown can be:
Application detect issues and start closing external processes inside the container ( services, binaries etc)- in case those services/binaries are unresponsive, kill -9 will be used, then just "rm -f" of the container.

**Comment by thaJeztah** [comment]
@gmistick in that case it does look like it's working as expected; If (for example) processes in the container are logging to a file (instead of stdout/stderr) in the container's filesystem, then those directories can grow rapidly. If you're building a lot of images that don't share layers, those take up space too

**Comment by gmistick** [comment]
But im not increasing containers count.
Im removing and re creating.

**Comment by thaJeztah** [comment]
> But im not increasing containers count.
> Im removing and re creating.

Are you only starting containers, or also building images (`docker build`); are you removing dangling images in between?

**Comment by gmistick** [comment]
Images were built and pushed to the registry 2-3 resets ago. ( i built last
version 1 month ago, i had to reset var/lib/docker 3 times already )

So im not building so much images...
Just creating and removing , creating and removing

**Comment by gmistick** [comment]
I'm now using Overlay2, and i still saw increase in disk usage.
docker ps -s showed me that some of the containers reached a size of 50G.. (Its not the logs because they are written to a mapped binded volume ).

Any advice on how to debug what piece of the container is taking up that size?
I can say that the containers that take a lot of space are ones that had a lot of restarts.

**Comment by cpuguy83** [comment]
@gmistick `docker diff <container id>` wil show you the changes in the container FS.

**Comment by jadametz** [comment]
Seeing the same. Spent most of last Friday in the Docker community Slack trying to get some answers on this. This is a very real issue blocking my organization from efficiently building Docker images. Please let me know what information I can provide to help.

**Comment by gmistick** [comment]
0. Are you restarting containers a lot?
1. Docker ps -a -s will tell you size of each container.  Make sure its not
the containers that eat up your space.
2. I moved to.overlayfs2 as it was old recommendation from the above
comments.

**Comment by PhilMarkham** [comment]
I've hit this issue 3 times in the last 6 or so months on a server that has relatively-high container churn. At any one time we usually have maybe 30-60 containers running across 5-10 different docker-compose stacks representing different applications. Throughout the day we probably see around 10 deployments on average. All of these applications mount host directories via the 'volumes' directive in the compose yaml files.


The deployment process for each app stack is along the lines of;
docker-compose up, a few docker-compose run calls & docker-compose stop on the old stack

The run calls do operations that are quite FS-intensive; the tools to build each app are wrapped up in a container and they therefore do things like 'npm install' to build the app into the volume-mounted directories.

So we don't generally do anything that kills anything or should leave anything an untidy state. But, to help that end, we do periodically (daily) also tidy up containers & images with;
```
docker rm $(docker ps -f status=exited -q);
docker rmi $(docker images -q -f dangling=true);
```
And also remove any old deployment directories (which were previously mounted as volumes via docker-compose)


Despite this, inode usage continues to creep to 100% and the only way I've found to 'fix' the issue when it happens is;
stop all docker containers
remove all containers & images 
stop the docker service
rm -fr /var/lib/docker
rm -fr /path/to/all/deployment/volumes
start docker service
re-deploy all our applications


Which is obviously not that great of a process, especially when aufs is considered 'production ready' in the docker docs.

I'm currently looking into other storage drivers; overlay1 has similar inode issues, overlay2 isn't supported by ubuntu 14.04's kernel version, btrfs & zfs aren't generally considered production ready - so I'm left with device mapper LVM; but after a quick google, I see a few issues floating around and mixed views as to whether they've actually been fixed.


I'm very happy to work with you guys to help figure out the cause and a fix for this issue. But in the meantime; What is the recommended, most-stable storage driver for production use? 



My current set up details:
Docker server version: 1.12.1
Storage driver: aufs
Backing filesystem: extfs
OS: Ubuntu 14.04.5 LTS

**Comment by DanielSchwiperich** [comment]
I came here because we ran out of inodes on one of our gitlab ci runners. Solution was to clear unwanted volumens which existet even the container were long gone.

`docker volume rm $(docker volume ls -qf dangling=true)`

Credits to source: http://blog.yohanliyanage.com/2015/05/docker-clean-up-after-yourself/

**Comment by cpuguy83** [comment]
@DanielSchwiperich You might try updating your scripts to use `docker rm -v` when removing containers so that it will also clean up any anonymous volumes.

1.13 has a new `docker system df` and `docker system prune` command to help clean up potentially unused resources.

**Comment by Hubbitus** [comment]
@PhilMarkham RHEL/CentOs by default use devmapper but recommend BTRFS. I also use BTRFS on pair of servers without of problems in few months. Including gitlab-ci builders and run tests in containers (so workload is same: up->run tests->stop and remove)

**Comment by mixalturek** [comment]
Hi, I was facing 100% of allocated inodes in /var/lib/docker/overlay and found this issue by searching the web. There is a file system resource leak somewhere among docker, kernel and overlay fs that is triggered by parallel execution of containers. The bug is very simple to reproduce, minimal scenario is below.

Terminal window 1: Watch inodes allocations (replace /var by partition where your /var/lib/docker resides), monitoring only.

```
watch -d -n 1 "df -i | grep /var"
```

Terminal window 2 - N: Fire docker containers in a loop and watch the inodes allocation. Docker cleans everything properly if there is only one loop running (no parallelism). The resource leak occurs only with parallel execution, more simultaneous loops makes the issue more likely to occur. It's common threading or timing issue which is typically very hard to debug.

```
for i in {0..99}
do
	docker run --rm alpine:3.5 echo $i Hello world!
done
```

Quick observations

- Docker doesn't delete all created files and directories as it should when `docker --rm` is used.
- Docker doesn't see the leaked files (`docker ps -a`, `docker volume ls -q -f dangling=true` and similar API calls), so the only possibility is to fully erase /var/lib/docker and start again.
- When looking to /var/lib/docker/overlay the bug produces mostly empty directories or empty files, but the files have a content sometimes.
- Different "garbage" directories contain different sets of directories/files, the pattern seems random which supports the theory about threading issue.
- Issue is reproducible on CentOS 7, default kernel 3.10.0, Docker 17.03.1-ce, storage driver overlay.
- Issue is NOT reproducible on CentOS 7, kernel 4.9.20, Docker 17.03.1-ce, storage driver overlay2 (this is a known workaround).

**Comment by dmcgowan** [comment]
@mixalturek  it is expected that running containers using `overlay` in parallel will use significantly more inodes. I am more concerned about your claim that files are not deleted after the completion of `docker run --rm`. After all containers have finished running (assuming you can get the number down a point where it doesn't error out), is there still higher inode usage with the containers removed? If the containers are failing to be removed, then I would expect some indication in the daemon logs as to why the removal failed. If the containers are reported as removed but inode usage is still high, can you also look for failure log messages for an indication as to why the on disk resources were not cleaned up. If you see these log messages please share as we are always trying to find scenarios that might help us fix races on resource cleanup.

**Comment by mixalturek** [comment]
@dmcgowan Yes, you are understanding my report correctly. The leaked files persist in /var/lib/docker/overlay after all containers exit (with zero/success status), Docker cleans them using `docker run --rm` and `docker ps -a` shows empty list. I found the errors below in /var/lib/messages, I expect someone accidentally holds a file descriptor so the umount fails. Tho errors are fired only when the containers are started/stopped in parallel, logs with serial execution are clean.

```
Apr 25 08:57:00 luft-docker-prod-idc-prg020 dockerd: time="2017-04-25T08:57:00.483996030+02:00" level=error msg="Error removing mounted layer 9e5aec31e653d7264bd8362c4292c976b66fe0328d53ab5ef7ad2e4f6f28910b: remove /var/lib/docker/overlay/3902986bff7a5b67417aca457efc6dfd7d76ea8e7b276653b734423fcaa59612/merged: device or resource busy"
Apr 25 08:57:00 luft-docker-prod-idc-prg020 dockerd: time="2017-04-25T08:57:00.484277918+02:00" level=error msg="can't remove container 9e5aec31e653d7264bd8362c4292c976b66fe0328d53ab5ef7ad2e4f6f28910b: Driver overlay failed to remove root filesystem 9e5aec31e653d7264bd8362c4292c976b66fe0328d53ab5ef7ad2e4f6f28910b: remove /var/lib/docker/overlay/3902986bff7a5b67417aca457efc6dfd7d76ea8e7b276653b734423fcaa59612/merged: device or resource busy"
...
Apr 25 08:57:00 luft-docker-prod-idc-prg020 dockerd: time="2017-04-25T08:57:00.949118847+02:00" level=error msg="can't remove container efd3ad2ef8174972af5ff93c87d7626ed1805516e6ff5c6d407c3743295a8646: Unable to remove filesystem for efd3ad2ef8174972af5ff93c87d7626ed1805516e6ff5c6d407c3743295a8646: remove /var/lib/docker/containers/efd3ad2ef8174972af5ff93c87d7626ed1805516e6ff5c6d407c3743295a8646/shm: device or resource busy"
...
Apr 25 08:57:01 luft-docker-prod-idc-prg020 dockerd: time="2017-04-25T08:57:01.140351214+02:00" level=error msg="Error removing mounted layer 515c96bbb2319d09caac425a4fb38221934c3f17ec55895ca31b560c50e13204: remove /var/lib/docker/overlay/c4c37ef942120cc1b0337cab40904bd0f86647555aac34a15ba4488d73069420/merged: device or resource busy"
Apr 25 08:57:01 luft-docker-prod-idc-prg020 dockerd: time="2017-04-25T08:57:01.140445520+02:00" level=error msg="can't remove container 515c96bbb2319d09caac425a4fb38221934c3f17ec55895ca31b560c50e13204: Driver overlay failed to remove root filesystem 515c96bbb2319d09caac425a4fb38221934c3f17ec55895ca31b560c50e13204: remove /var/lib/docker/overlay/c4c37ef942120cc1b0337cab40904bd0f86647555aac34a15ba4488d73069420/merged: device or resource busy"
...
```

```
[root@luft-docker-prod-idc-prg020 ~]# find /var/lib/docker/overlay/3902986bff7a5b67417aca457efc6dfd7d76ea8e7b276653b734423fcaa59612
/var/lib/docker/overlay/3902986bff7a5b67417aca457efc6dfd7d76ea8e7b276653b734423fcaa59612
/var/lib/docker/overlay/3902986bff7a5b67417aca457efc6dfd7d76ea8e7b276653b734423fcaa59612/merged
[root@luft-docker-prod-idc-prg020 ~]# find /var/lib/docker/containers/efd3ad2ef8174972af5ff93c87d7626ed1805516e6ff5c6d407c3743295a8646
/var/lib/docker/containers/efd3ad2ef8174972af5ff93c87d7626ed1805516e6ff5c6d407c3743295a8646
/var/lib/docker/containers/efd3ad2ef8174972af5ff93c87d7626ed1805516e6ff5c6d407c3743295a8646/shm
[root@luft-docker-prod-idc-prg020 ~]# find /var/lib/docker/overlay/c4c37ef942120cc1b0337cab40904bd0f86647555aac34a15ba4488d73069420
/var/lib/docker/overlay/c4c37ef942120cc1b0337cab40904bd0f86647555aac34a15ba4488d73069420
/var/lib/docker/overlay/c4c37ef942120cc1b0337cab40904bd0f86647555aac34a15ba4488d73069420/merged
[root@luft-docker-prod-idc-prg020 ~]# 
```

```
[root@luft-docker-prod-idc-prg020 ~]# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
[root@luft-docker-prod-idc-prg020 ~]# df -i | grep /var
/dev/mapper/system-var      655360 88125 567235   14% /var
[root@luft-docker-prod-idc-prg020 ~]# ls -l /var/lib/docker/overlay/ | wc -l
5920
[root@luft-docker-prod-idc-prg020 ~]#
```

**Comment by mixalturek** [comment]
Workaround for this issue is probably `MountFlags=private` in systemd for docker daemon. It's possible that it has been fixed in a recent version of Docker. Related issues are below.

https://github.com/moby/moby/issues/22260
https://github.com/moby/moby/issues/25718#issuecomment-250356570

**Comment by anthony-o** [comment]
I'm using `overlay2` and it seems to be worst, even after pruning the system, and running 5 containers, here is the inodes usage in `/var/lib/docker`:
```
sudo du --max-depth=1 --inodes /var/lib/docker/
51      /var/lib/docker/containers
12324287        /var/lib/docker/overlay2
1       /var/lib/docker/tmp
588     /var/lib/docker/image
2       /var/lib/docker/builder
3       /var/lib/docker/network
1       /var/lib/docker/trust
5       /var/lib/docker/plugins
2       /var/lib/docker/volumes
1       /var/lib/docker/swarm
12324942        /var/lib/docker/
```
Here is my `docker info`:
```
Containers: 5
 Running: 5
 Paused: 0
 Stopped: 0
Images: 41
Server Version: 17.09.0-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 06b9cb35161009dcb7123345749fef02f7cea8e0
runc version: 3f2f8b84a77f73d38244dd690525642a72156c64
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-98-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 6.65GiB
# [...]
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Comment by cpuguy83** [comment]
@anthony-o I suspect this is because docker is doing a lazy unmount+delete to prevent `device or resource busy` errors.
Unfortunately this means that instead of `device or resource busy` you have stuff that appears to be removed but it's not really because it exists in another mount namespace.

I believe https://github.com/moby/moby/pull/36047 and https://github.com/moby/moby/pull/36055 should fully such cases.

**Comment by cpuguy83** [comment]
I'm going to close this as it should be fully resolved in 17.12.1.
If you are still seeing this problem *after* upgrading to 17.12.1 please open a new issue.

Thanks!
